### PR TITLE
reliable consumer

### PR DESCRIPTION
**Pull request by Gsantomaggio**
Signed-off-by: Gabriele Santomaggio <G.santomaggio@gmail.com>

The second part of https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/issues/92

- Implement ReliableConsumer 
- Refactor the code by adding the ReliableBase class that's common for Producer and Consumer 
- Add an example for the Consumer

How to test:

```csharp
using System.Buffers;
using System.Net;
using System.Net.Security;
using System.Text;
using RabbitMQ.Stream.Client;
using System.Text.Json;
using RabbitMQ.Stream.Client.Reliable;

namespace example;

public class Reliable
{
    public class Conn
    {
        public string name { get; set; }
        public Dictionary<string, string> client_properties { get; set; }
    }

    private async Task<int> HttpKillConnections()
    {
        using var handler = new HttpClientHandler
        {
            Credentials = new NetworkCredential("guest", "guest"),
        };
        using var client = new HttpClient(handler);
        var result = await client.GetAsync("http://localhost:15672/api/connections");
        var json = await result.Content.ReadAsStringAsync();
        var connections = JsonSerializer.Deserialize<IEnumerable<Conn>>(json);
        if (connections == null) return 0;
        foreach (var conn in connections)
        {
            await client.DeleteAsync($"http://localhost:15672/api/connections/{conn.name}");
        }

        return connections.Count();
    }


    public async Task RProducerConsumer()
    {
        Console.WriteLine("Reliable .NET Producer/Cosumer");
        var config = new StreamSystemConfig();
        const string stream = "my-reliable-stream-consumer";
        var system = await StreamSystem.Create(config);
        try
        {
            await system.DeleteStream(stream);
        }
        catch (Exception)
        {
            // ignored
        }

        await system.CreateStream(new StreamSpec(stream)
        {
        });
        const int totalMessages = 22_111_111;
        var countConfirm = 0;
        var countError = 0;
        var run = Task.Run(async () =>
        {
            var reliableProducer = await ReliableProducer.CreateReliableProducer(new ReliableProducerConfig()
            {
                StreamSystem = system,
                Stream = stream,
                Reference = "my-reliable-producer",
                ClientProvidedName = "my-reliable-producer",
                ConfirmationHandler = confirmation =>
                {
                    if (confirmation.Status != ConfirmationStatus.Confirmed)
                    {
                        countError++;
                    }
                    else
                    {
                        countConfirm++;
                    }


                    return Task.CompletedTask;
                }
            });


            Task.Run(() =>
            {
                while (reliableProducer.IsOpen())
                {
                    Console.WriteLine(
                        $"Confirmed: {countConfirm} error: {countError} - TOTAL: {countConfirm + countError}");
                    Thread.Sleep(2000);
                }
            });
            for (var i = 0; i < totalMessages; i++)
            {
                await reliableProducer.Send(new Message(Encoding.UTF8.GetBytes($"hello {i}")));
            }
        });

        var consumed = 0;
        var rConsumer = await ReliableConsumer.CreateReliableConsumer(new ReliableConsumerConfig()
        {
            Stream = stream,
            StreamSystem = system,
            OffsetSpec = new OffsetTypeFirst(),
            ClientProvidedName = "My-R-Consumer",
            MessageHandler = async (consumer, ctx, message) =>
            {
                consumed++;
                await Task.CompletedTask;
            }
        });

        Task.Run(() =>
        {
            while (rConsumer.IsOpen())
            {
                Console.WriteLine(
                    $"Total consumed messages: {consumed}");
                Thread.Sleep(2000);
            }
        });

        for (var i = 0; i < 40; i++)
        {
            Thread.Sleep(5000);
            try
            {
                Console.WriteLine($"Killing all the connections...");
                var r = HttpKillConnections();
                r.Wait();
            }
            catch (Exception e)
            {
                Console.WriteLine($"fail to kill {e}");
            }
        }

        run.Wait();
    }


}
```

What I would expect is to have:
```
Confirmed: 21498967 error: 612144 - TOTAL: 22111111 ## <---- eguals to the total messages number 
Total consumed messages: 21508657 ## <---- Messages count on the queue 
```

### CODE REVIEW

**Comment by Gsantomaggio** [comment]
@simone-fariselli  if you have the possibility can you please test it?

**Comment by simone-fariselli** [comment]
Hi @Gsantomaggio,
I do some tests and it looks ok.

File: RabbitMQ.Stream.Client/Reliable/IReconnectStrategy.cs [hunk 1]
```diff
@@ -2,10 +2,32 @@
...
 
+using System;
+using System.Threading;
+
 namespace RabbitMQ.Stream.Client.Reliable;
 
 public interface IReconnectStrategy
 {
-    void WhenDisconnected(out bool reconnect);
+    void WhenDisconnected(out bool reconnect, string connectionInfo);
```
[the remark this answers is not in the document]
Comment (Gsantomaggio) [reply RabbitMQ.Stream.Client/Reliable/IReconnectStrategy.cs:12]: Will change.

[the remark this answers is not in the document]
Comment (Gsantomaggio) [reply RabbitMQ.Stream.Client/Reliable/IReconnectStrategy.cs]: Closed via https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/pull/115/commits/8de01623d9b00206e4eacc922872bee477f312cd

**Pull request merged**